src/realm/list.cpp

```diff
@@ -408,8 +408,12 @@ Mixed Lst<Mixed>::set(size_t ndx, Mixed value)
     if (Replication* repl = Base::get_replication()) {
         repl->list_set(*this, ndx, value);
     }
-    if (!(old.is_same_type(value) && old == value)) {
+    if (!value.is_same_type(old) || value != old) {
         do_set(ndx, value);
+        if (value.is_type(type_Dictionary, type_List)) {
+            m_tree->ensure_keys();
+            set_key(*m_tree, ndx);
```

Comment:
did this work properly before? 🤔

Reply:
You're right. I guess in some cases the key would not be updated.